nemo/collections/tts/models/melgan.py

```diff
@@ -0,0 +1,320 @@
+                    return_list[i].append(loss)
+            for i, loss in enumerate(return_list):
+                return_list[i] = torch.mean(torch.stack(loss))
+            return return_list
+
+        loss = torch.stack([x['loss'] for x in outputs]).mean()
+        self.log("val_loss", loss, sync_dist=True)
+
+        if self.start_training_disc:
+            gan_loss = get_stack(outputs, "gan_loss")
+            self.log("val_loss_gan_loss", sum(gan_loss) / len(gan_loss), sync_dist=True)
+            for i, _ in enumerate(gan_loss):
+                self.log(
+                    f"val_loss_gan_loss_{i}", gan_loss[i] / len(gan_loss), sync_dist=True,
+                )
+
+        sc_loss = get_stack(outputs, "sc_loss")
+        mag_loss = get_stack(outputs, "mag_loss")
+        self.log("val_loss_feat_loss", torch.stack([x['loss_feat'] for x in outputs]).mean(), sync_dist=True)
+        self.log("val_loss_feat_loss_fb_sc", sum(sc_loss) / len(sc_loss), sync_dist=True)
+        self.log("val_loss_feat_loss_fb_mag", sum(mag_loss) / len(sc_loss), sync_dist=True)
+        for i, _ in enumerate(sc_loss):
+            self.log(f"val_loss_feat_loss_fb_sc_{i}", sc_loss[i] / len(sc_loss), sync_dist=True)
+            self.log(f"val_loss_feat_loss_fb_mag_{i}", mag_loss[i] / len(sc_loss), sync_dist=True)
+
+    def __setup_dataloader_from_config(self, cfg, shuffle_should_be: bool = True, name: str = "train"):
+        if "dataset" not in cfg or not isinstance(cfg.dataset, DictConfig):
+            raise ValueError(f"No dataset for {name}")
+        if "dataloader_params" not in cfg or not isinstance(cfg.dataloader_params, DictConfig):
+            raise ValueError(f"No dataloder_params for {name}")
+        if shuffle_should_be:
+            if 'shuffle' not in cfg.dataloader_params:
+                logging.warning(
+                    f"Shuffle should be set to True for {self}'s {name} dataloader but was not found in its "
+                    "config. Manually setting to True"
+                )
+                with open_dict(cfg["dataloader_params"]):
+                    cfg.dataloader_params.shuffle = True
+            elif not cfg.dataloader_params.shuffle:
+                logging.error(f"The {name} dataloader for {self} has shuffle set to False!!!")
+        elif not shuffle_should_be and cfg.dataloader_params.shuffle:
+            logging.error(f"The {name} dataloader for {self} has shuffle set to True!!!")
+
+        dataset = instantiate(cfg.dataset)
+        return torch.utils.data.DataLoader(dataset, collate_fn=dataset.collate_fn, **cfg.dataloader_params)
+
+    def setup_training_data(self, cfg):
+        self._train_dl = self.__setup_dataloader_from_config(cfg)
+
+    def setup_validation_data(self, cfg):
+        self._validation_dl = self.__setup_dataloader_from_config(cfg, shuffle_should_be=False, name="validation")
+
+    def training_epoch_end(self, outputs):
+        # Do manual logginging of learning rate and epoch
```

Comment:
Why manual?

Reply:
I could never get lightning's self.log to work properly, so I reverted to tensorboard's API